docs(routing): add a routing guide

vvo opened the pull request:
This PR includes:
- the `routing` guide, [review it live](https://deploy-preview-2879--algolia-instantsearch.netlify.com/v2/guides/routing.html)
- deprecation of the urlSync option via console.warn
- minor change to CONTRIBUTING for dev on examples
- routing for all examples (simple one)
- minor fix for no-results guide

vvo commented:
One thing I wanted to achieve is to easily handle the case of already existing querystring parameters.

For example let's say I load https://website.com/?loggedIn=1 and I want to use the routing feature. What I'd like to do is:

```js
const search = instantsearch({
  // ... appId, apiKey...
  routing: {
    stateMapping: {
      stateToRoute(uiState) {
        return {
		  loggedIn: window.isUserLoggedIn
          q: uiState.query
		};
      },
      routeToState(routeState) {
        return {
          query: routeState.q
        };
      }
    }
  }
});
```

This ^ works and will add loggedIn query parameter to the url. But what if window.isUserLoggedIn state changes from the outside (user logs out, it's an ajax call)? I would like to turn the url to: `https://website.com/?loggedIn=0&q=phone`.

To do so for now I guess I would have to create my own router or own widget?

What I may want as a user:

```js
const search = instantsearch(...);
search.router.write(); // triggers write on the history router
```

There are implications to this so I just want to have your opinion on the proposal.

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+your search UI (which widget were refined, what is the current search query ..) with any kind of storage.
+And most probably you want that storage to be the browser url bar.
+
+Synchronizing your UI with the browser url is a good practice. It allows any of your user to take one of your result page, copy paste the
```
your users

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+
+This guide provides you examples for the most common use cases along with some advanced ones. You can also read the [routing API reference](instantsearch.html#struct-RoutingOptions).
+
+As you will see with advanced examples, creating your own url scheme to map your needs requires some good coding. But keep in mind
```
a significant amount of code, but

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+
+`https://website.com/?query=a&refinementList%5Bbrand%5D%5B0%5D=Drama`
+
+While
```
While not (one line for legibility of the source)

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+
+`https://website.com/?query=a&brands=Sony~Samsung&page=2`
+
+This way your users to be able to more easily read them when shared via emails, documents, social media...
```
This way your users will be able to read them more easily when

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+
+To do so, the `routing` option accepts a simple boolean but also more complex objects to allow for customization. The first customization option
+you want to use is `stateMapping`. It allows you to define more precisely how the state of your search will be synchronized to your
+browser bar (or any other router storage you might have).
```
URL bar

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+  // ... appId, apiKey...
+  routing: {
+    router: instantsearch.routers.history({
+      windowTitle: function(routeState) {
```
we can use either all modern features (object shorthand) or none (no template strings either) so that people can copy-paste it in sites without build process

vvo replied:
I want to keep "routeState" here for the sake of explaining which kind of objects you are receiving. Users can then write that however they want. Works?

Haroenv replied:
I meant:

```diff
- windowTitle: function(routeState) {
+ windowTitle(routeState) {
```

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+Using that we can configure:
+- `windowTitle`: This method can be used to map the object (`routeState`) returned from `stateToRoute` to your window title
+- `createURL`: This method is called everytime we need to create a url. When we want to synchronize the `routeState` to the browser url bar,
+when we want to render `<a href>` tags in the `menu` widget or when you call `createURL` in one of your connectors rendering method
```
`connectors'`

Haroenv commented on src/lib/InstantSearch.js:
```diff
@@ -82,10 +82,28 @@ Usage: instantsearch({
+      }
+
+      console.warn(
+        'For advanced usages docs, check out https://community.algolia.com/instantsearch.js/v2/guides/routing.html#migrating-from-urlsync'
```
usage documentation

bobylito commented on src/lib/InstantSearch.js:
```diff
@@ -82,10 +82,28 @@ Usage: instantsearch({
 
-    if (urlSync && routing) {
-      throw new Error(
-        'InstantSearch configuration error: it is not possible to use `urlSync` and `routing` at the same time'
```
I would still keep the error because the behavior is completely undefined.

vvo replied:
The error is right here but below (after the warnings)

vvo replied:
I will put it on top again it's better

bobylito replied:
Sorry, missed it 👍

Haroenv commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,276 @@
+And most probably you want that storage to be the browser url bar.
+
+Synchronizing your UI with the browser url is a good practice. It allows any of your user to take one of your result page, copy paste the
+browser url and send it to a friend.
```
I wouldn't mention it, because it's tricky to say if we handled it correctly (is it correct to have a new history entry for every refinement or not)

vvo replied:
Oh yes but Ill definitely mention that you can use back/next buttons and reload without any issue, yes!

Haroenv commented on docgen/src/guides/routing.md:
````diff
@@ -0,0 +1,276 @@
+});
+```
+
+There's a lifecycle dictating when the `stateMapping` functions are called:
````
There's a lifecycle in which the `stateMapping` functions are called

clear enough IMo

vvo commented:
Thanks a lot to you two for the extensive review. I implemented every change you suggested, they were on point.

I also removed the super weird wrapping I had, sorry about that. I now have configured vscode the right way to handle it.

bobylito commented:
🚀

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+- `stateToRoute` is called whenever widgets are refined (clicked). It is also called everytime any widget needs to create a URL.
+- `routeToState` is called whenever the user loads, reloads the page or click on back/next buttons of the browser.
+
+To build your own mapping easily, just `console.log(uiState)` and see what you're getting. Note that the object you return in `stateToRoute` will be the one you'll receive as an argument in `routeToState`.
```
Why not give them the actual object format?

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+Using that we can configure:
+- `windowTitle`: This method can be used to map the object (`routeState`) returned from `stateToRoute` to your window title
+- `createURL`: This method is called everytime we need to create a url. When we want to synchronize the `routeState` to the browser url bar, when we want to render `<a href>` tags in the `menu` widget or when you call `createURL` in one of your connectors's rendering method
+- `parseURL`: This method is called everytime the user loads, reloads the page or click on back/next buttons of the browser
```
`click` → `clicks`

francoischalifour commented on docgen/src/guides/routing.md:
````diff
@@ -0,0 +1,258 @@
+Allow: *
+```
+
+Now that you know how to create such advanced url synchronization mechanism, the only next step would be to create your own router. But let's first have a look at the full `uiState` reference.
````
`only next` → `final`?

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+
+## `instantsearch.routers.history` API
+
+InstantSearch.js provides a default router under `instantsearch.routers.history`. You can use it when you want to go futher than just aliasing querystring parameters in the url. For example if you want to generate urls like `https://website.com/search/q/phone/brands/Sony~Samsung/p/1`.
```
`querystring parameters` → `query string parameters`

(also change references below if you decide to change it here)

We should probably add a link to the [definition of a Query String](https://en.wikipedia.org/wiki/Query_string) since this is the first time we talk about it in this guide.

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+This function allows you to directly change the format of urls that will be created and rendered to the browser url bar or widgets.
+This function is called everytime InstantSearch.js needs to create a URL. **The provided options** are:
+  - *`qsModule`: object*, a querystring parsing and stringifying module, [full documentation](https://github.com/ljharb/qs). We use it internally so we provide it to you as a convenience
+  - *`location`: function*, alias to window.location
```
"window.location" → "`window.location`" (wrap with `code` tag)

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+
+- **`opts.parseURL: function({qsModule, location})`**
+This function is responsible for parsing back the url string to a `routeState`. This function must be customized if you customized the `createURL` function.
+This function is called everytime the user loads, reloads or click on back/next buttons of the browser. **The provided options** are:
```
`click` → `clicks`

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+This function is responsible for parsing back the url string to a `routeState`. This function must be customized if you customized the `createURL` function.
+This function is called everytime the user loads, reloads or click on back/next buttons of the browser. **The provided options** are:
+  - *`qsModule`: object*, a querystring parsing and stringifying module, [full documentation](https://github.com/ljharb/qs). We use it internally so we provide it to you as a convenience.
+  - *`location`: function*, alias to window.location.
```
"window.location" → "`window.location`" (wrap with `code` tag)

francoischalifour commented on docgen/src/guides/routing.md:
```diff
@@ -0,0 +1,258 @@
+You must return an `object`. Which is a `routeState`.
+
+- **`opts.writeDelay: number, default 400`**
+This option controls the number of milliseconds to wait before actually writing the new url to the browse urlr bar. You can think about it this way:
```
`urlr` → `URL`

bobylito commented:
> There are implications to this so I just want to have your opinion on the proposal.

I've had the same sort of requirement. This is too awkward that you have to trigger the `change` yourself on the Helper.

Haroenv commented:
Agree, would be nice to call write in those scenarios
